### PR TITLE
fix(server): Fix server tests failing intermittently due to `ioredis`

### PR DESCRIPTION
**Pull request by ThatOneBro**
A few contributors have reported flakiness of tests in `packages/server/src/app.test.ts`, specifically since using real Redis in tests. There were a few things that indicated it may be to weird interplay between `jest` and `redis`, although the details are unclear. I tweaked a few things that makes it feel more stable (haven't had another test failure yet) but not 100% confident this is the fix.

### CODE REVIEW

File: packages/server/src/app.ts [hunk 1]
```diff
@@ -214,7 +214,9 @@ export async function shutdownApp(): Promise<void> {
   closeRateLimiter();
 
   if (server) {
-    server.close();
+    await new Promise((resolve) => {
+      (server as ReturnType<typeof http.createServer>).close(resolve);
```
Comment (codyebberson) [inline packages/server/src/app.ts:218]:
```suggestion
      (server as http.Server).close(resolve);
```

**Review by codyebberson — APPROVED** [review]

File: packages/server/src/websockets.ts [hunk 2]
```diff
@@ -34,12 +41,34 @@ export function initWebSockets(server: http.Server): void {
     // See: https://github.com/websockets/ws/blob/master/doc/ws.md#websocketbinarytype
     socket.binaryType = 'nodebuffer';
 
+    if (!wsState?.sockets.size) {
+      let socketsClosedResolve!: () => void;
+      const socketsClosedPromise = new Promise<void>((resolve) => {
+        socketsClosedResolve = resolve;
+      });
+      wsState = { sockets: new Set(), socketsClosedPromise, socketsClosedResolve };
```
Comment (codyebberson) [inline packages/server/src/websockets.ts:49]: Tricky

**Review by codyebberson — APPROVED** [review]
lgtm 👍 

This only affects tests & server shutdown, right?  Looks safe.

**Comment by ThatOneBro** [comment]
> lgtm 👍
> 
> This only affects tests & server shutdown, right? Looks safe.

Yeah this is only regarding tests and server shutdown. Thanks 🙏

**Pull request merged**